lexer.py

```diff
@@ -0,0 +1,124 @@
+from tokens import TokenTypes, Token
+
+ESCCHARS = {
+    "n":  "\n",
+    "t":  "\t",
+    "\\": "\\",
+    "'": "'",
+    "r":  "\r",
+}
+
+RESERVED_WORDS = {
+    "SELECT",
+    "FROM",
+    "EDGES",
+    "NODES",
+    "TRAVERSE",
+    "BY",
+    "DEPTH",
+    "BREADTH",
+    "WHERE",
+    "LIMIT",
+}
+
+class LexingError(Exception):
+    pass
+
+class Lexer:
+    def __init__(self, text):
+        self.text = text         # string input
+        self.pos = 0             # self.pos is index into self.text
+        self.line = 1            # current line
+        self.current_char = self.text[0] # current char
+        self.tokens = []         # tokens so far
+    def _error(self, msg):
+        raise LexingError("LexingError, line " + str(self.line) + ": " + msg + ".")
```

Comment:
No need to include `LexingError` in the message; the repr for it will include "LexingError: " as the first part.  Also the Python convention for exception messages is to generally not end with a period.